ostd/src/mm/page/mod.rs

```diff
@@ -52,24 +53,36 @@ impl<M: PageMeta> Page<M> {
     ///
     /// The function panics if:
     ///  - the physical address is out of bound or not aligned;
-    ///  - the page is already in use.
+    ///  - the page is already in use;
+    ///  - the metadata's size exceeds [`PAGE_METADATA_MAX_SIZE`];
+    ///  - the alignment of the metadata exceeds [`PAGE_METADATA_MAX_ALIGN`].
     pub fn from_unused(paddr: Paddr, metadata: M) -> Self {
         assert!(paddr % PAGE_SIZE == 0);
         assert!(paddr < MAX_PADDR.load(Ordering::Relaxed) as Paddr);
+
+        // Checking unsafe preconditions of the `PageMeta` trait.
+        assert!(size_of::<M>() <= PAGE_METADATA_MAX_SIZE);
+        assert!(align_of::<M>() <= PAGE_METADATA_MAX_ALIGN);
+
         let vaddr = mapping::page_to_meta::<PagingConsts>(paddr);
         let ptr = vaddr as *const MetaSlot;
 
-        // SAFETY: The aligned pointer points to an initialized `MetaSlot`.
-        let usage = unsafe { &(*ptr).usage };
-        // SAFETY: The aligned pointer points to an initialized `MetaSlot`.
+        // SAFETY: The aligned pointer points to a initialized `MetaSlot`.
         let ref_count = unsafe { &(*ptr).ref_count };
 
-        usage
-            .compare_exchange(0, M::USAGE as u8, Ordering::SeqCst, Ordering::Relaxed)
-            .expect("page already in use when trying to get a new handle");
+        ref_count
+            .compare_exchange(0, 1, Ordering::Acquire, Ordering::Relaxed)
+            .expect("Page already in use when trying to get a new handle");
+
+        // SAFETY: The aligned pointer points to a initialized `MetaSlot`.
+        let vtable_ptr = unsafe { (*ptr).vtable_ptr.get() };
 
-        let old_ref_count = ref_count.fetch_add(1, Ordering::Relaxed);
-        debug_assert_eq!(old_ref_count, 0);
+        // SAFETY: The pointer is valid and we have the exclusive access.
+        unsafe {
+            vtable_ptr.write(core::ptr::metadata(
+                &metadata as &dyn PageMeta as *const dyn PageMeta,
```

Comment:
`as *const dyn PageMeta` can be removed because it's an implicit conversion and won't harm the code readability.